Architectural documentation (#12)
Energinet-DataHub/energy-origin: doc/architecture/README.md

# Platform Architecture Specification

This document contains the overall architecture of the EnergyOrigin platform

## Intended readers and confidentiality

This document is intended for Architects and Developers
using and building the platform.

This document is public as it is a part of the OpenSource project
called EnergyOrigin.
  
---

## Architecture Overview

This chapter describes the background of our architecture, its purposes,
constraints and high level architecture.

### Motivation

As part of the Green transition in the energy sector Granular Certificates are
becoming a central part of the puzzle to solve the ever growing need to be able
to document the green and/or renewable aspect of energy.

> The changes in how energy is handled and consumed is changed in a increasing

Reviewer: ```suggestion
The changes in how energy is handled and consumed is changed in an increasing
```

> The changes in how energy is handled and consumed is changing in an increasing
> manner, so a platform that can handle these challenges are needed to enable the
> green transition.

Reviewer 2: ```suggestion
The changes in how energy is handled and consumed is changing in an increasing
manner, so a platform that can handle these challenges is needed to enable the
green transition.
```

### Architectural Goals and Constraints

> The goal of the architecture is to:

Reviewer 2: ```suggestion
The goals of the architecture are:
```

- **Platform**: create a extendable platform where individual domains are
loosely coupled to enabled domains to be added, replaced or removed with

> - **Isolated Domains**: the platform isolate groups of services into domains,
> each domain can have larger responsibility than a single service
> **no changed** required to other domains.

Reviewer: ```suggestion
- **Isolated Domains**: the platform isolates groups of services into domains,
each domain can have a larger responsibility than a single service.
**No changes** required to other domains.
```

### System Business Requirements and Constraints

Business requirements:

- **OpenSource**: the platform must be OpenSource and a community created around it.

> This will enabled a shared benefit over time.

Reviewer: ```suggestion
This will enable a shared benefit over time.
```

are at the core of the platform.
The design is to store data as events in a shared Event Store shared between domains.
Individual domains can then be added, updated and removed with low coupling.

A new domain can be created and brought up to speed based on the Event Store.
This removes the need for inter domain calls to pull and migrate data from other domains when domains are added or updated.

---

## Test Strategy

All code must be sufficiently tested by unit and integrations tests,
so that the developers feel safe when deploying without any gates to production.

All manual tests will be regarded as tech-dept.

---

## Architecture overview

The diagrams below are documented based on [arc42](https://docs.arc42.org/section-5/)

### High Level context diagram

Context diagram, high level overview of the system that will be part of the Danish deployment.

The blue is the platform itself.
Green are systems that has been integrated to.
Yellow are expected future integrations.

![Overview of the different systems integrating](diagrams/context.drawio.png)

## EnergyOrigin internals

> This view explores how the different components that the platform is build from
> integrates with each other.

Reviewer 2: ```suggestion
This view explores how the different components that the platform is built from
integrates with each other.
```

![Exploded view of EnergyOrigin](diagrams/level1.drawio.png)

## Domains

Below is a list of the domains and a reference to their documentation.

*TODO: document missing domains.*

> - [Measurement](domains/measurements.md)

Reviewer 2: This and the linked document seems irrelevant with the addition of the `docs/api` area, can this (and the linked document) be deleted?
